### PR TITLE
feat(connectors): net.dns_lookup — full dig parity via dnspython (#2409)

### LINKED ISSUE #2409: Task: net.dns_lookup — full dig parity (typed records + chosen resolver) via dnspython, forward/reverse
## Summary

`net.dns_lookup(name, type?, resolver?)` on Initiative #2405's T1 scaffolding: full `dig` parity from the backplane's vantage — forward records of any type (A/AAAA/CNAME/MX/TXT/SRV/NS/SOA/PTR), reverse PTR, and an optional **chosen resolver** so an operator can compare "what the pod's resolver returns" vs "what an authoritative/other nameserver returns" (the split-horizon case that forced registering a vCenter by IP). Adds `dnspython` — approved.

## Current state (verified in code)

- The only resolver in-tree is `socket.getaddrinfo` via `asyncio.to_thread` (`ssrf_guard.py:224-245`) — forward A/AAAA + reverse PTR only, system resolver only. No typed records, no resolver selection. No `dnspython` dep yet.

## Desired state

- Add `dnspython` to `backend/pyproject.toml` + `uv.lock`.
- `net.dns_lookup` handler (T1 mold: targetless, return-failures contract, audit-visible name+type+resolver): `dns.asyncresolver.Resolver`; `type` param (default `A`) drives `resolve(name, rdtype)`; a `resolver` param (IP) sets `.nameservers=[resolver]` to query a chosen server; reverse form when `name` is an IP (`dns.reversename.from_address` → PTR). Report `{name, type, resolver: "system"|<ip>, records: [{type, value, ttl}], authoritative?}`.
- `NXDOMAIN` / `NoAnswer` / `SERVFAIL` / timeout → `{"resolved": false, "reason": "nxdomain"|"no_answer"|"servfail"|"timeout"}` with `status="ok"`.
- Gate the queried `name` (and, for a custom `resolver`, the resolver IP) through T1's `assert_probe_allowed` — one guard applied consistently (#1177); querying an internal resolver / resolving internal names is itself mild recon.

## Acceptance criteria

- [ ] Returns A, AAAA, CNAME, MX, TXT, SRV, NS, SOA for the corresponding test names, and reverse PTR for an IP — targetless, fresh boot.
- [ ] A `resolver=<ip>` param queries that nameserver (pinned against a fixture resolver returning a distinct answer from the system resolver — the split-horizon proof).
- [ ] NXDOMAIN / failure returns `{resolved: false, reason}` with `status="ok"` — never a `connector_*` error.
- [ ] The queried name + type + resolver are recorded in the audit row; `name` and any custom `resolver` IP are probe-allowlist-gated.
- [ ] `dnspython` added to pyproject + lock; scans green.

## Out of scope

- DNSSEC chain validation (report the `AD` flag if present, but don't validate). Zone transfers (AXFR). Recursive trace from the root (that's `dig +trace` territory — file separately only if requested).

## References

- Parent: #2405. Depends-on: T1. `dns.asyncresolver` / `dns.reversename`. Prior resolver primitive `ssrf_guard.py:224-245`.
- Best-practices — *python*: async resolver off the event loop; failure is the product (`status=ok`); one guard applied uniformly (#1177).

Parent goal: #221
Parent initiative: #2405

### PR DESCRIPTION
**Pull request by damir-topic**
## Summary

- Adds `net.dns_lookup` on the T1 `net.*` mold (#2406): a synthetic, **targetless/class-less** DNS diagnostic that resolves from the backplane's vantage via **dnspython** (`dns.asyncresolver`, off the event loop). Gives operators `dig` parity where the only in-tree resolver (`socket.getaddrinfo`) offered forward A/AAAA + reverse PTR against the system resolver alone.
- **Forward** typed records (A/AAAA/CNAME/MX/TXT/SRV/NS/SOA, enum-bounded so metaqueries/AXFR stay unreachable at the param gate); **reverse PTR** when `name` is an IP literal (`dig -x`); an optional **`resolver` IP** pins `nameservers` so an operator can compare the pod resolver vs an authoritative/other nameserver (the split-horizon case).
- **One guard, uniformly (#1177):** the queried `name` **and** any custom `resolver` IP are screened through T1's `assert_probe_allowed` before any query; a non-IP resolver is rejected up front (`bad_resolver`).
- **Return-failures contract:** NXDOMAIN / no-answer / SERVFAIL / timeout / no-resolver-config → `{resolved: false, reason}` with `status="ok"`, never a `connector_*` error. `name`/`type`/`resolver` are audit-visible via `raw_payload`. AA reported as `authoritative`; the DNSSEC AD flag reported (not validated — chain validation / AXFR / `+trace` out of scope).
- `dnspython` (ISC-licensed) pinned as a **direct** dependency — already present transitively via `email-validator`/`pymongo`; the pin makes the connector's runtime requirement explicit and passes the Python License Check allow-list.

## Test plan

- [x] `ruff check` / `ruff format --check` — clean on changed files
- [x] `mypy src/meho_backplane/connectors/net/ops.py` — clean
- [x] code-quality diff gate — 0 blockers
- [x] `pytest tests/test_connectors_net_dns_lookup.py` — 18 passed (real dnspython wire path against an in-process UDP fixture resolver; rootless, offline)
- [x] `pytest tests/test_connectors_net_probe.py` — 17 passed (registrar now upserts both `net.*` ops)
- [x] broadcast/classify/typed_register subset — 208 passed
- [x] AC: forward A/AAAA/CNAME/MX/TXT/SRV/NS/SOA + reverse PTR (parametrized, targetless, fresh boot)
- [x] AC: `resolver=<ip>` split-horizon proof — two fixtures on `127.0.0.1`/`127.0.0.2` sharing one port return distinct answers
- [x] AC: NXDOMAIN / no_answer / servfail / timeout → `resolved:false` + reason, `status="ok"`
- [x] AC: audit row `raw_payload` carries `name`/`type`/`resolver`; `name` + custom `resolver` allowlist-gated
- [x] AC: `dnspython` added to pyproject + lock (single 2.8.0 entry, direct dep line added)

## Out of scope

- DNSSEC chain validation (AD flag reported, not validated), zone transfers (AXFR), recursive `dig +trace` — per issue.
- Sibling `net.*` T2/T3 ops (`tls_inspect` / `http_probe`) — separate tasks under Initiative #2405; they extend the same `ops.py` registrar (orchestrator serializes merges).

Closes #2409

### CODE REVIEW

**Comment by damir-topic** [comment]
@coderabbitai review

File: backend/src/meho_backplane/connectors/net/ops.py [hunk 1]
```diff
@@ -253,18 +262,349 @@ async def net_tcp_check(operator: Operator, target: Any, params: dict[str, Any])
...
+    name = str(params["name"]).strip()
+    requested_type = str(params.get("type") or "A").upper()
+    raw_resolver = params.get("resolver")
+    resolver_ip = str(raw_resolver).strip() if raw_resolver not in (None, "") else None
+    timeout = _clamp_timeout(params.get("timeout_seconds", _DEFAULT_TIMEOUT_SECONDS))
+    resolver_label = resolver_ip or "system"
+
+    # A chosen resolver must be an IP literal: dnspython's ``nameservers``
+    # setter rejects a hostname, and an unresolved name pins no server.
+    if resolver_ip is not None and _as_ip_literal(resolver_ip) is None:
```
Comment (damir-topic) [inline backend/src/meho_backplane/connectors/net/ops.py]: **[🛑 B1]** — Bracketed IPv6 `resolver` (`[::1]`) escapes the return-failures contract with an uncaught `ValueError`

**What:** `_as_ip_literal` strips brackets and accepts `[::1]` (its docstring: "so a caller can pass either shape"), and `assert_probe_allowed` also strips brackets — so a bracketed IPv6 resolver passes both the `bad_resolver` gate and the allowlist gate. But the un-stripped `resolver_ip` is then handed to `_build_resolver` → `resolver.nameservers = [resolver_ip]`, and dnspython's `nameservers` setter rejects the bracketed form.

**Why it matters:** The op's core contract is "a failed lookup is the product, never a `connector_*` error." A bracketed IPv6 resolver — a shape the code explicitly advertises accepting — instead raises `ValueError: nameserver [::1] is not … an IP address …`, which surfaces as a connector error. Reproduced through `net_dns_lookup(...)`: `RAISED: ValueError`. (Bare `::1` works; only the bracketed form crashes.)

**Required change:** After the `bad_resolver` check, normalize to the bare literal before use, e.g.:
```python
if resolver_ip is not None:
    literal = _as_ip_literal(resolver_ip)
    if literal is None:
        return _dns_refusal(name, requested_type, resolver_label, "bad_resolver")
    resolver_ip = literal
    resolver_label = literal
```
so both `resolver.nameservers` and the audit `resolver` label are bracket-free. Add a `resolver="[::1]"` test asserting a clean result (not a raised error).

**Comment by damir-topic** [comment]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #2409 — `net.dns_lookup`: full `dig` parity (typed records + chosen resolver) via dnspython
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `74502df1`
**AI reviewer (CodeRabbit):** not received — rate-limited ("Review limit reached, next review available in 32 minutes"); Claude verdict posted per policy

### Acceptance criteria
- [x] Forward A/AAAA/CNAME/MX/TXT/SRV/NS/SOA + reverse PTR — targetless, fresh boot (parametrized tests, real dnspython wire path against in-process UDP fixture; 18 passed).
- [x] `resolver=<ip>` queries that nameserver — split-horizon proof (two fixtures on 127.0.0.1/127.0.0.2 sharing a port return distinct answers).
- [x] NXDOMAIN / no_answer / SERVFAIL / timeout → `{resolved:false, reason}` with `status="ok"`, never `connector_*` — **for the tested inputs.** See B1: one untested input shape (bracketed IPv6 resolver) still escapes this contract.
- [x] Audit `raw_payload` carries name/type/resolver; `name` + custom `resolver` IP both allowlist-gated (verified: name verbatim, resolver by range; non-IP resolver → `bad_resolver`; unlisted resolver → `not_in_probe_allowlist`).
- [x] `dnspython>=2.7,<3` added to pyproject + `uv.lock`; Python License Check green (ISC).

### Predicted vs. actual diff
- ✅ `connectors/net/ops.py` — handler, param/response schemas, LLM instructions, second registrar upsert
- ✅ `tests/test_connectors_net_dns_lookup.py` — real-wire fixture resolver, split-horizon, failure contract, audit, allowlist
- ✅ `backend/pyproject.toml` + `backend/uv.lock` — `dnspython` direct pin
- ✅ `CHANGELOG.md` — distinct `### Added — net.dns_lookup …` header (no [Unreleased] merge collision)
- ✅ `docs/codebase/connectors-net-diagnostics.md` — T4 section added
- No undisclosed scope, no missing predicted file. No FastAPI route touched → no OpenAPI snapshot regen needed (CLI API snapshot freshness check green).

### Integration verification
- [x] `ruff check` — clean (ops.py + test)
- [x] `ruff format --check` — clean
- [x] `mypy src/.../net/ops.py` — clean
- [x] `pytest tests/test_connectors_net_dns_lookup.py` — 18 passed (ran in worktree)
- [x] Split-horizon / return-failures / allowlist / audit paths exercised end-to-end via the real dispatch path
- n/a migrations (no migration; "Python (database migrations)" CI lane SKIPPED)

### Findings (inline — each is a merge blocker per team policy)
- 🛑 **B1** — Bracketed IPv6 `resolver` (`[::1]`) raises an uncaught `ValueError` → `connector_*` error, breaking the return-failures contract. `_as_ip_literal` accepts the bracketed form (its docstring says "so a caller can pass either shape") and the allowlist strips brackets, so `[::1]` passes both the `bad_resolver` gate and `assert_probe_allowed` — but the un-stripped `resolver_ip` is then handed to `resolver.nameservers = [resolver_ip]`, whose dnspython setter rejects the bracketed form (`nameserver [::1] is not … an IP address …`). Reproduced through the handler: `RAISED: ValueError`. Bare `::1` works; only the advertised bracketed shape crashes. **Required change:** after the `bad_resolver` check, normalize to the bare literal, e.g. `resolver_ip = _as_ip_literal(resolver_ip)` (and derive `resolver_label` from it) so the value passed to `nameservers` — and the audit label — are bracket-free. Add a test with `resolver="[::1]"`.

### Adjacent improvements (non-blocking, not posted inline)
- `ops.py` is now 666 lines, over the 600 soft block-limit (WARN only in diff-mode; CodeRabbit may also flag it). A per-op split is **premature now** — it would churn the shared `register_net_typed_operations` registrar for marginal gain — but with T2/T3 (`tls_inspect`/`http_probe`, #2405) landing in the same file it will cross ~1000 lines. Recommend deferring the split until T2 lands, then extracting per-op modules (`ops/dns_lookup.py`, `ops/tcp_check.py`, `ops/_shared.py` for `_clamp_timeout`/timeout constants + a thin registrar). Not a blocker for this PR.
- `dns.resolver.NoNameservers` → `servfail` also covers an allowlisted-but-unreachable custom resolver; the `servfail` label is a slight over-generalization but acceptable for a diagnostic. No change required.

### AI reviewer (CodeRabbit) cross-check
- CodeRabbit was rate-limited and produced no findings this iteration (next slot ~32 min). Verdict is Claude-only; the orchestrator should note CR did not review this PR.

### Docs
- [x] `docs/codebase/connectors-net-diagnostics.md` — updated with the T4 `net.dns_lookup` section (gating, return-failures, DNSSEC-reported-not-validated, dependency note). Accurate against the code.

### What worked well (review-body only)
- The security-sensitive part is right: name gated **verbatim** (no resolve-to-allow rebind window), custom resolver gated by range, `search=False` blocks search-domain expansion — split-horizon resolver-pinning cannot widen the allowlisted probe space. The enum bound on `type` keeps `ANY`/`AXFR` unreachable at the param gate. DNSSEC AD is faithfully "reported, not validated" in schema, docstring, and comment. Tests drive the real dnspython wire path against an in-process fixture rather than mocking the resolver.

**Comment by damir-topic** [comment]
@coderabbitai review

**Review by damir-topic — COMMENTED** [review]
## /auto-review-pr summary (iteration 2)




**Anchor task:** #2409 — net.dns_lookup (full dig parity via dnspython)
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** APPROVE
**Head SHA reviewed:** `50d6d7df`
**AI reviewer (CodeRabbit):** review finished, incremental — no new findings on the fix commit (it does not re-review already-reviewed commits)

### B1 (iteration-1 blocker) — resolved

The bracketed-IPv6 resolver crash is genuinely fixed. Traced end-to-end:

- `resolver="[::1]"` → `_as_ip_literal("[::1]")` returns the bare `::1` → `resolver_ip` and `resolver_label` are both reassigned to `::1` (ops.py:548-557) **before** the allowlist gate and `_build_resolver`.
- `assert_probe_allowed("::1")` runs on the bare literal; `_build_resolver` then sets `resolver.nameservers = ["::1"]` — the bare form dnspython accepts (empirically confirmed: bare `::1` accepted, `[::1]` raises `ValueError`).
- The audit label records the normalized bare `::1` (line 557 → returned `resolver` field).
- Result is a clean `{resolved: false, reason: "timeout", resolver: "::1"}` — no `connector_*` error, return-failures contract intact.

The added `test_bracketed_ipv6_resolver_is_normalized` exercises the **real** `nameservers` setter: it stubs only `Resolver.resolve` (to raise `Timeout`), not `_build_resolver`, so the setter still runs on `["::1"]`. Had the brackets not been stripped, the setter would raise before `resolve` is reached. Asserts `resolved is False`, `reason == "timeout"`, `resolver == "::1"`.

**Port-suffix edge (`[::1]:5353`) confirmed rejected upstream:** `_as_ip_literal("[::1]:5353")` returns `None` (does not end with `]`, so brackets aren't stripped and `ipaddress.ip_address` raises) → `bad_resolver` refusal. Empirically verified.

### Acceptance criteria
- [x] Bracketed IPv6 resolver `[::1]` returns a clean structured result, not a raised `connector_*` error — new test + trace
- [x] return-failures contract intact — full 19-test suite green (NXDOMAIN / no_answer / servfail / timeout / bad_resolver / allowlist paths)
- [x] resolver port edge case confirmed rejected — `[::1]:5353` → `bad_resolver`

### Regression sweep (iteration-1 non-blocking properties still hold)
- [x] Allowlist gates **both** the queried name and the custom resolver IP (ops.py:563-565)
- [x] Record-type enum bounded; `ANY`/`AXFR` unreachable at the param-schema gate (ops.py:266-296)
- [x] Native-async: `dns.asyncresolver` + `await resolver.resolve` — no event-loop blocking
- [x] dnspython 2.8.0 license ISC — clean
- [x] Return-failures contract: all failure reasons mapped, none raised
- [x] Main-merge (#2420 keycloak seam, `50d6d7df`) did **not** touch net.* code — only the B1 fix commit `19388326` modified `connectors/net/`

### Integration verification
- [x] `ruff check` — clean
- [x] `ruff format --check` — 2 files already formatted
- [x] `mypy ops.py` — no issues
- [x] `pytest tests/test_connectors_net_dns_lookup.py` — 19 passed (real dnspython wire path, incl. new normalization test)
- [x] Empirical repro: `_as_ip_literal` normalization + dnspython setter accept/reject behavior confirmed in the worktree venv
- ⚠️ not run: docker compose / agent eval (not relevant to a synthetic typed op)

### Findings (inline)
None. The iteration-1 B1 inline thread is now outdated (the fixed line moved).

### Adjacent improvements (non-blocking, not posted inline)
- `ops.py` remains over the 600-line soft block-limit (WARN only). Per-op split stays deferred until T2/T3 (tls_inspect / http_probe) land in the same file — unchanged accepted recommendation, not re-litigated here.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 0 — CodeRabbit ran incrementally and produced no new findings on the fix commit.
- Disputed: 0
- Out of scope: 0

### Docs
- [x] `docs/codebase/connectors-net-diagnostics.md` — updated in the original PR; unchanged by the fix and still accurate.

### What worked well
- The fix is exactly the minimal normalization the iteration-1 blocker called for, placed before both the allowlist gate and the resolver build so the bare literal is used consistently for gating, the `nameservers` setter, and the audit label.
- The new test deliberately stubs only `resolve` so the real `nameservers` setter — the actual crash site — is exercised, rather than mocking around it.
